### PR TITLE
Make search experiment by tag key comparison case sensitive in MYSQL

### PR DESCRIPTION
**Pull request by WeichenXu123**
Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

## Related Issues/PRs

#xxx

## What changes are proposed in this pull request?

Make search experiment by tag key comparison case sensitive in MYSQL

## How is this patch tested?

- [ ] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/store/tracking/sqlalchemy_store.py [hunk 1]
```diff
@@ -1363,7 +1363,9 @@ def _get_search_experiments_filter_clauses(parsed_filters, dialect):
             val_filter = SearchUtils.get_sql_filter_ops(
                 SqlExperimentTag.value, comparator, dialect
             )(value)
-            key_filter = SqlExperimentTag.key == key
+            key_filter = SearchUtils.get_sql_filter_ops(
+                SqlExperimentTag.key, "=", dialect
+            )(key)
```
Comment (harupy) [inline mlflow/store/tracking/sqlalchemy_store.py:1368]: Can you add a test?

Comment (WeichenXu123) [reply mlflow/store/tracking/sqlalchemy_store.py:1368]: added.

File: tests/store/tracking/test_sqlalchemy_store.py [hunk 2]
```diff
@@ -502,6 +502,20 @@ def test_search_experiments_pagination(self):
...
+        from mlflow.store.db.db_types import MYSQL
+        filter_str = f"tag.abc = 'xyz'"
+        parsed_filter = SearchExperimentsUtils.parse_search_filter(filter_str)
+
+        with mock.patch(
+                "mlflow.utils.search_utils.SearchUtils.get_sql_filter_ops",
+                return_value=SqlExperimentTag.key.__eq__
+        ) as mock_get_sql_filter_ops:
+            _get_search_experiments_filter_clauses(parsed_filter, MYSQL)
+            mock_get_sql_filter_ops.assert_called_with(SqlExperimentTag.key, "=", MYSQL)
```
Comment (harupy) [inline tests/store/tracking/test_sqlalchemy_store.py:517]: We don't need to use mock. We can update `test_search_experiments_filter_by_tag`.

Comment (harupy) [reply tests/store/tracking/test_sqlalchemy_store.py:517]: We can call `search_experiments` with `filter_string="tag.KEY = 'value'"` and check the result is empty.

Comment (harupy) [reply tests/store/tracking/test_sqlalchemy_store.py:517]: Not sure if you know this, but we run tests against all the supported databases. See https://github.com/mlflow/mlflow/tree/master/tests/db

Comment (WeichenXu123) [reply tests/store/tracking/test_sqlalchemy_store.py:517]: cool.

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @WeichenXu123 !

**Comment by harupy** [comment]
In MySQL, `BINARY` might not work for a primary key.

**Comment by WeichenXu123** [comment]
> In MySQL, `BINARY` might not work for a primary key.
> 
> ```
>         experiments = self.store.search_experiments(filter_string="tag.KEY = 'value'")
> >       assert len(experiments) == 0
> E       AssertionError: assert 1 == 0
> E        +  where 1 = len([<Experiment: artifact_location='artifact_folder/1', experiment_id='1', lifecycle_stage='active', name='exp1', tags={'key': 'value'}>])
> ```
> 
> If it's challenging to fix this issue, we should document it as a limitation for MySQL. I can help.

what about remove "key" column from the primary-key ?
we have "experiment_id" as  primary-key so that it won't hurt performance too much.

File: mlflow/utils/search_utils.py [hunk 3]
```diff
@@ -105,15 +105,23 @@ def get_sql_filter_ops(cls, column, operator, dialect):
         if dialect == MSSQL:
             column = column.collate("Japanese_Bushu_Kakusu_100_CS_AS_KS_WS")
 
+        # Use non-binary ahead of binary comparison for runtime performance
+        col = f"{column.class_.__tablename__}.{column.key}"
         # Use non-binary ahead of binary comparison for runtime performance
         def case_sensitive_mysql_eq(value):
-            return sa.and_(column.__eq__(value), column.op("= BINARY", is_comparison=True)(value))
+            return sa.text(f"{col} = :value AND BINARY {col} = :value").bindparams(
```
Comment (WeichenXu123) [inline mlflow/utils/search_utils.py:113]: So this is because `column.op("= BINARY")` does not work ?

Comment (harupy) [reply mlflow/utils/search_utils.py:113]: yes

**Review by harupy — APPROVED** [review]

**Pull request merged**